File: ruby/calculator_cli_app/src/application.rb
# rbs_inline: enabled

require_relative './queries/calculation_query'
require_relative './validations/args_validation'

module CalculatorCliApp
  class Application
    include ::Validations::ArgsValidation

    # @rbs args: Array[String]
    # @rbs return: void
    def initialize(args)
      @args_size = args.length
      @seed      = args.first
      @n         = args.last
    end

    # @rbs return: void
    def run
      validate!(args_size, seed, n.to_i.nonzero?)

      result = calculation_query.f(n)
      puts result
    end

    private

    attr_reader :args_size, :seed, :n

    # @rbs return: Queries::CalculationQuery
    def calculation_query
      @calculation_query ||= ::Queries::CalculationQuery.new(seed)
    end
  end
end

File: javascript/calculator_cli_app/src/application.js
import { CalculationQuery } from './queies/calculation_query';
import { validate } from './validations/args_validation';

const Application = class {
  constructor(args) {
    this.args_size        = args.length;
    this.seed             = args[0];
    this.n                = args[args.length -1];
    this.calculation_query = null;
  }

  run() {
    validate(this.args_size, this.seed, parseInt(this.n));

    const result = this.__calculation_query__().f(parseInt(this.n));
    console.log(result);
  }

  // private

  __calculation_query__() {
    if (this.__calculation_query__ === null) {
      this.__calculation_query__ = new CalculationQuery(this.seed);
      return this.__calculation_query__;
    } else {
      return this.__calculation_query__;
    }
  }
}

export {
  Application
};
